File: leetcode/0114-flatten-binary-tree-to-linked-list/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        node = root

        while node:
            if node.left:
                # find the rightmost node of the left subtree
                rightmost = node.left
                while rightmost.right:
                    rightmost = rightmost.right

                # graft the original right subtree onto that rightmost node
                rightmost.right = node.right

                # the left subtree becomes the new right subtree, left is cleared
                node.right = node.left
                node.left = None

            node = node.right

File: leetcode/0114-flatten-binary-tree-to-linked-list/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        TreeNode node = root;

        while (node != null) {
            if (node.left != null) {
                // find the rightmost node of the left subtree
                TreeNode rightmost = node.left;
                while (rightmost.right != null) {
                    rightmost = rightmost.right;
                }

                // graft the original right subtree onto that rightmost node
                rightmost.right = node.right;

                // the left subtree becomes the new right subtree, left is cleared
                node.right = node.left;
                node.left = null;
            }

            node = node.right;
        }
    }
}
